{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare Meta Data for Tryptase Expressing vs Non-Tryptase Expressing\n",
    "\n",
    "1) tryptase-expressing v. non-expressing – you would likely have to set some somewhat arbitrary RPKM threshold based upon controls/confidence\n",
    "\n",
    "2) among tryptase-positive: alphaDUP-expressing v. all alpha-expressing v. non-alpha-expressing\n",
    "\n",
    "3) if anything comes out of 2, it might make sense to iterate this back through step 1 (comparing alphaDUP v. non-expressing for instance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('../data_files/mbase_normed/aml.meta.mbase_norm.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay_Type</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Center_Name</th>\n",
       "      <th>DATASTORE_filetype</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>InsertSize</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>...</th>\n",
       "      <th>alpha_read_covers_snps_count_exact</th>\n",
       "      <th>alpha_dup_read_covers_snps_count_exact</th>\n",
       "      <th>beta_read_covers_snps_count_exact</th>\n",
       "      <th>norm_alignment_count</th>\n",
       "      <th>norm_alpha_wt_count</th>\n",
       "      <th>norm_alpha_dup_count</th>\n",
       "      <th>norm_beta_count</th>\n",
       "      <th>norm_alpha_read_covers_snps_count_exact</th>\n",
       "      <th>norm_alpha_dup_read_covers_snps_count_exact</th>\n",
       "      <th>norm_beta_read_covers_snps_count_exact</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR330925</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>PRJNA145991</td>\n",
       "      <td>SAMN00709303</td>\n",
       "      <td>GEO</td>\n",
       "      <td>sra</td>\n",
       "      <td>SRX091763</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR330926</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>PRJNA145991</td>\n",
       "      <td>SAMN00709304</td>\n",
       "      <td>GEO</td>\n",
       "      <td>sra</td>\n",
       "      <td>SRX091764</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR330927</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>PRJNA145991</td>\n",
       "      <td>SAMN00709305</td>\n",
       "      <td>GEO</td>\n",
       "      <td>sra</td>\n",
       "      <td>SRX091765</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.034965</td>\n",
       "      <td>0.013986</td>\n",
       "      <td>0.013986</td>\n",
       "      <td>0.006993</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR330928</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>PRJNA145991</td>\n",
       "      <td>SAMN00709306</td>\n",
       "      <td>GEO</td>\n",
       "      <td>sra</td>\n",
       "      <td>SRX091766</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005450</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR330929</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36</td>\n",
       "      <td>PRJNA145991</td>\n",
       "      <td>SAMN00709307</td>\n",
       "      <td>GEO</td>\n",
       "      <td>sra</td>\n",
       "      <td>SRX091767</td>\n",
       "      <td>0</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Assay_Type AssemblyName  AvgSpotLen   BioProject     BioSample  \\\n",
       "SRR330925    RNA-Seq          NaN          36  PRJNA145991  SAMN00709303   \n",
       "SRR330926    RNA-Seq          NaN          36  PRJNA145991  SAMN00709304   \n",
       "SRR330927    RNA-Seq          NaN          36  PRJNA145991  SAMN00709305   \n",
       "SRR330928    RNA-Seq          NaN          36  PRJNA145991  SAMN00709306   \n",
       "SRR330929    RNA-Seq          NaN          36  PRJNA145991  SAMN00709307   \n",
       "\n",
       "          Center_Name DATASTORE_filetype Experiment  InsertSize  \\\n",
       "SRR330925         GEO                sra  SRX091763           0   \n",
       "SRR330926         GEO                sra  SRX091764           0   \n",
       "SRR330927         GEO                sra  SRX091765           0   \n",
       "SRR330928         GEO                sra  SRX091766           0   \n",
       "SRR330929         GEO                sra  SRX091767           0   \n",
       "\n",
       "                            Instrument                  ...                    \\\n",
       "SRR330925  Illumina Genome Analyzer II                  ...                     \n",
       "SRR330926  Illumina Genome Analyzer II                  ...                     \n",
       "SRR330927  Illumina Genome Analyzer II                  ...                     \n",
       "SRR330928  Illumina Genome Analyzer II                  ...                     \n",
       "SRR330929  Illumina Genome Analyzer II                  ...                     \n",
       "\n",
       "          alpha_read_covers_snps_count_exact  \\\n",
       "SRR330925                                  0   \n",
       "SRR330926                                  0   \n",
       "SRR330927                                  0   \n",
       "SRR330928                                  0   \n",
       "SRR330929                                  0   \n",
       "\n",
       "          alpha_dup_read_covers_snps_count_exact  \\\n",
       "SRR330925                                      0   \n",
       "SRR330926                                      0   \n",
       "SRR330927                                      0   \n",
       "SRR330928                                      0   \n",
       "SRR330929                                      0   \n",
       "\n",
       "          beta_read_covers_snps_count_exact norm_alignment_count  \\\n",
       "SRR330925                                 0             0.000000   \n",
       "SRR330926                                 0             0.000000   \n",
       "SRR330927                                 0             0.034965   \n",
       "SRR330928                                 0             0.005450   \n",
       "SRR330929                                 0             0.000000   \n",
       "\n",
       "           norm_alpha_wt_count  norm_alpha_dup_count norm_beta_count  \\\n",
       "SRR330925             0.000000              0.000000        0.000000   \n",
       "SRR330926             0.000000              0.000000        0.000000   \n",
       "SRR330927             0.013986              0.013986        0.006993   \n",
       "SRR330928             0.000000              0.000000        0.005450   \n",
       "SRR330929             0.000000              0.000000        0.000000   \n",
       "\n",
       "          norm_alpha_read_covers_snps_count_exact  \\\n",
       "SRR330925                                     0.0   \n",
       "SRR330926                                     0.0   \n",
       "SRR330927                                     0.0   \n",
       "SRR330928                                     0.0   \n",
       "SRR330929                                     0.0   \n",
       "\n",
       "          norm_alpha_dup_read_covers_snps_count_exact  \\\n",
       "SRR330925                                         0.0   \n",
       "SRR330926                                         0.0   \n",
       "SRR330927                                         0.0   \n",
       "SRR330928                                         0.0   \n",
       "SRR330929                                         0.0   \n",
       "\n",
       "          norm_beta_read_covers_snps_count_exact  \n",
       "SRR330925                                    0.0  \n",
       "SRR330926                                    0.0  \n",
       "SRR330927                                    0.0  \n",
       "SRR330928                                    0.0  \n",
       "SRR330929                                    0.0  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_dup(df):\n",
    "    \"\"\"\n",
    "    Does the sample have any reads covering the snps that characterise the dup locus?\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    if df['norm_alpha_dup_read_covers_snps_count_exact'] == 0:\n",
    "        \n",
    "        return False\n",
    "    else:\n",
    "        \n",
    "        return True\n",
    "    \n",
    "def has_tryptase()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['has_dup'] = df.apply(has_dup, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.000000     111\n",
       "0.003937       5\n",
       "0.002625       5\n",
       "0.001312       4\n",
       "0.001351       3\n",
       "0.009186       3\n",
       "0.002668       2\n",
       "0.029289       2\n",
       "0.016216       2\n",
       "0.024557       2\n",
       "0.043330       2\n",
       "0.001658       2\n",
       "0.002561       2\n",
       "0.136364       2\n",
       "0.020833       2\n",
       "0.003649       1\n",
       "0.769476       1\n",
       "0.024295       1\n",
       "0.002807       1\n",
       "0.603706       1\n",
       "0.237988       1\n",
       "0.268441       1\n",
       "0.013686       1\n",
       "0.242475       1\n",
       "0.016262       1\n",
       "0.034875       1\n",
       "0.125607       1\n",
       "0.671599       1\n",
       "0.075885       1\n",
       "0.003125       1\n",
       "            ... \n",
       "0.812070       1\n",
       "0.007208       1\n",
       "0.003368       1\n",
       "0.561208       1\n",
       "0.004647       1\n",
       "0.013189       1\n",
       "0.015756       1\n",
       "0.621366       1\n",
       "0.026144       1\n",
       "18.869549      1\n",
       "0.062935       1\n",
       "0.034346       1\n",
       "2.196636       1\n",
       "1.117069       1\n",
       "0.018075       1\n",
       "0.011550       1\n",
       "1.026758       1\n",
       "0.574143       1\n",
       "0.739311       1\n",
       "0.009770       1\n",
       "1.233262       1\n",
       "0.339782       1\n",
       "0.115210       1\n",
       "0.019129       1\n",
       "0.023219       1\n",
       "0.317204       1\n",
       "0.007442       1\n",
       "0.765539       1\n",
       "0.542817       1\n",
       "0.016946       1\n",
       "Name: norm_alignment_count, Length: 2591, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['norm_alignment_count'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
